### PR TITLE
meta/sql: retry txn for mariadb when conflict

### LINKED ISSUE #5442: pjdfstest failed for mariadb
**What happened**:
https://github.com/juicedata/juicefs/actions/runs/12495811510

```
Test Summary Report
-------------------
tests/chown/00.t          (Wstat: 0 Tests: 1323 Failed: 0)
  TODO passed:   1097, 1101, 1107, 1112, 1116, 1122, 1127
                1131, 1137, 1142, 1146, 1152, 1157, 1161
                1167, 1172, 1176, 1182, 1187
tests/mkfifo/03.t         (Wstat: 256 Tests: 4 Failed: 0)
  Non-zero exit status: 1
Files=236, Tests=8819, 199 wallclock secs ( 0.86 usr  0.31 sys + 33.89 cusr  9.84 csys = 44.90 CPU)
Result: FAIL
```

client logs：
```
2024/12/25 19:05:21.246428 juicefs[2350] <INFO>: JuiceFS version 5.1.6 (2024-12-13 ec7a5ec) [mount.go:634]
2024/12/25 19:05:21.639239 juicefs[2350] <INFO>: Reduce cache size to 65 GiB [disk_cache.go:1292]
2024/12/25 19:05:21.639276 juicefs[2350] <INFO>: Cache: /var/jfsCache/ci-coverage capacity: 65 GiB [disk_cache.go:1295]
2024/12/25 19:05:21.639758 juicefs[2350] <INFO>: Mounting volume ci-coverage at /home/runner/work/juicefs/juicefs/cover ... [mount_unix.go:822]
2024/12/25 19:05:26.109591 juicefs[2341] <INFO>: watching /home/runner/work/juicefs/juicefs/cover, pid 2350 [mount_unix.go:133]
2024/12/25 19:06:07.532494 juicefs[3104] <INFO>: Meta address: ***(127.0.0.1)/test [NewClient@interface.go:511]
2024/12/25 19:06:07.914423 juicefs[3113] <INFO>: Meta address: ***(127.0.0.1)/test [NewClient@interface.go:511]
2024/12/25 19:06:07.917509 juicefs[3113] <INFO>: Data use file:///var/jfs/pics/ [mount@mount.go:639]
2024/12/25 19:06:07.917545 juicefs[3113] <INFO>: JuiceFS version 1.3.0-dev+2024-12-25.52625b4 [mount@mount.go:679]
2024/12/25 19:06:07.918191 juicefs[3113] <INFO>: Disk cache (/var/jfsCache/3d87426c-5e08-457a-bcf0-ebf4329506df/): capacity (100 GiB), free ratio 10%, used ratio - [space 76.2%, inode 11.7%], max pending pages [15](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:16) [newCacheStore@disk_cache.go:145]
2024/12/25 19:06:07.937958 juicefs[3113] <INFO>: Create session 1 OK with version: 1.3.0-dev+2024-12-25.52625b4 [NewSession@base.go:476]
2024/12/25 19:06:07.941011 juicefs[3113] <INFO>: Prometheus metrics listening on 127.0.0.1:9567 [exposeMetrics@mount.go:135]
2024/12/25 19:06:07.941465 juicefs[3113] <INFO>: Mounting volume pics at /tmp/jfs ... [mountMain@mount_unix.go:883]
2024-12-25 19:06:12.638701 I | Unimplemented opcode POLL
2024/12/25 19:06:12.639238 juicefs[3104] <INFO>: watching /tmp/jfs, pid 3113 [watchdog@mount_unix.go:[16](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:17)0]
2024/12/25 19:08:39.496916 juicefs[3113] <ERROR>: error: Error 1020 (HY000): Record has changed since last read in table 'jfs_dir_stats'
goroutine 50527 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.22.10/x64/src/runtime/debug/stack.go:24 +0x5e
github.com/juicedata/juicefs/pkg/meta.errno({0x408eb20, 0xc001c43668})
	/home/runner/work/juicefs/juicefs/pkg/meta/utils.go:125 +0xfc
github.com/juicedata/juicefs/pkg/meta.(*dbMeta).doSyncDirStat(0xc000d283c0, {0x40cc538?, 0xc001af0cc0?}, 0x486)
	/home/runner/work/juicefs/juicefs/pkg/meta/sql.go:2678 +0xe8
github.com/juicedata/juicefs/pkg/meta.(*baseMeta).parallelSyncDirStat.func1(0x486)
	/home/runner/work/juicefs/juicefs/pkg/meta/quota.go:99 +0x94
created by github.com/juicedata/juicefs/pkg/meta.(*baseMeta).parallelSyncDirStat in goroutine 148
	/home/runner/work/juicefs/juicefs/pkg/meta/quota.go:97 +0xc5 [errno@utils.go:125]
2024/12/25 19:08:39.496967 juicefs[3113] <WARNING>: sync dir stat for 1158: input/output error [parallelSyncDirStat@quota.go:101]
2024/12/25 19:08:40.509849 juicefs[3113] <WARNING>: remove dir usage of ino(1199): Error 1020 (HY000): Record has changed since last read in table 'jfs_dir_stats' [doRmdir@sql.go:[17](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:18)80]
2024/12/25 19:08:40.510176 juicefs[3113] <ERROR>: error: Error 1020 (HY000): Record has changed since last read in table 'jfs_dir_stats'
goroutine 38817 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.22.10/x64/src/runtime/debug/stack.go:24 +0x5e
github.com/juicedata/juicefs/pkg/meta.errno({0x408eb20, 0xc001e33560})
	/home/runner/work/juicefs/juicefs/pkg/meta/utils.go:125 +0xfc
github.com/juicedata/juicefs/pkg/meta.(*dbMeta).doRmdir(0xc000d283c0, {0x40cd348, 0xc00219c190}, 0x4ae, {0xc000fa4b80, 0x7f}, 0xc0026ecf28, {0x0, 0x0, 0x0})
	/home/runner/work/juicefs/juicefs/pkg/meta/sql.go:[18](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:19)10 +0x269
github.com/juicedata/juicefs/pkg/meta.(*baseMeta).Rmdir(0xc000754008, {0x40cd348, 0xc00219c190}, 0x4ae, {0xc000fa4b80, 0x7f}, {0x0, 0x0, 0x0})
	/home/runner/work/juicefs/juicefs/pkg/meta/base.go:1325 +0x402
github.com/juicedata/juicefs/pkg/vfs.(*VFS).Rmdir(0xc0016200d0, {0x40ced80?, 0xc002[19](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:20)c190?}, 0x4ae, {0xc000fa4b80, 0x7f})
	/home/runner/work/juicefs/juicefs/pkg/vfs/vfs.go:304 +0x182
github.com/juicedata/juicefs/pkg/fuse.(*fileSystem).Rmdir(0xc000f196[20](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:21), 0xc002bb1d28?, 0xc0016ae1b0, {0xc000fa4b80, 0x7f})
	/home/runner/work/juicefs/juicefs/pkg/fuse/fuse.go:150 +0xc5
github.com/hanwen/go-fuse/v2/fuse.doRmdir(0x0?, 0xc00102a908)
	/home/runner/go/pkg/mod/github.com/juicedata/go-fuse/v2@v2.1.1-0.20241120081[21](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:22)7-1d367e67855d/fuse/opcode.go:395 +0x42
github.com/hanwen/go-fuse/v2/fuse.init.1.func1(0x1a00000038?, 0x2?)
	/home/runner/go/pkg/mod/github.com/juicedata/go-fuse/v2@v2.1.1-0.20241120081217-1d367e67855d/fuse/opcode.go:812 +0x49
github.com/hanwen/go-fuse/v2/fuse.(*Server).handleRequest(0xc0007abc20, 0xc00102a908)
	/home/runner/go/pkg/mod/github.com/juicedata/go-fuse/v2@v2.1.1-0.20241120081217-1d367e67855d/fuse/server.go:742 +0x2f3
github.com/hanwen/go-fuse/v2/fuse.(*Server).loop(0xc0007abc20, 0x1)
	/home/runner/go/pkg/mod/github.com/juicedata/go-fuse/v2@v2.1.1-0.20[24](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:25)1120081217-1d367e67855d/fuse/server.go:715 +0x118
created by github.com/hanwen/go-fuse/v2/fuse.(*Server).readRequest in goroutine 183
	/home/runner/go/pkg/mod/github.com/juicedata/go-fuse/v2@v2.1.1-0.20241120081217-1d367e67855d/fuse/server.go:448 +0x6b1 [errno@utils.go:1[25](https://github.com/juicedata/juicefs/actions/runs/12495811510/job/34866918682#step:11:26)]
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**

**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version:
- Cloud provider or hardware configuration running JuiceFS:
- OS (e.g `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Object storage (cloud provider and region, or self maintained):
- Metadata engine info (version, cloud provider managed or self maintained):
- Network connectivity (JuiceFS to metadata engine, JuiceFS to object storage):
- Others:

### PR DESCRIPTION
**Pull request by jiefenghuang**
close #5442

Transaction should have an X lock and wait for it, but MariaDB’s behavior is somewhat inconsistent and it reported an error.

Case: removeDir may conflict on child node (dirStat)

### CODE REVIEW

File: pkg/meta/sql.go [hunk 1]
```diff
@@ -840,6 +840,9 @@ func (m *dbMeta) txn(f func(s *xorm.Session) error, inodes ...Ino) error {
 		if eno, ok := err.(syscall.Errno); ok && eno == 0 {
 			err = nil
 		}
+		if err != nil {
+			logger.Warnf("Transaction failed: %s", err)
```
Comment (davies) [inline pkg/meta/sql.go:844]: There will be too much warning

Comment (jiefenghuang) [reply pkg/meta/sql.go]: pr just for test

**Comment by jiefenghuang** [comment]
> The same problem: https://github.com/juicedata/juicefs/actions/runs/12497648861/job/34870503833

passed.
image

File: pkg/meta/sql.go [hunk 2]
```diff
@@ -800,6 +800,9 @@ func (m *dbMeta) shouldRetry(err error) bool {
 		logger.Warnf("transaction failed: %s, will retry it. please increase the max number of connections in your database, or use a connection pool.", msg)
 		return true
 	}
+	if strings.Contains(msg, "error 1020 (hy000)") {
```
Comment (SandyXSD) [inline pkg/meta/sql.go:803]: Better add explanation about this error. Is it a common error or mariadb specific?

**Pull request merged**